crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +286,148 @@ impl Report for PrintJson {
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_start =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(start.as_ref(), true).count() + 1;
+                let column_end =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(msg.typo, true).count() + column_start;
+                let line_num = msg.context.as_ref().map(|context| match context {
+                    Context::File(context) => context.line_num,
+                    _ => 1,
+                }).unwrap_or(1);
+                
+                if msg.corrections == typos::Status::Valid {
+                    return Ok(());
+                }
+
+                let message = match &msg.corrections {
+                    typos::Status::Invalid => {
+                        format!("`{}` is disallowed", msg.typo)
+                    }
+                    typos::Status::Corrections(corrections) => {
+                        format!("`{}` should be {}", msg.typo, itertools::join(
+                            corrections.iter().map(|s| format!("`{}`", s)),
+                            ", ",
+                        ))
+                    }
+                    typos::Status::Valid => { unreachable!() }
+                };
+
+                let location = sarif::LocationBuilder::default()
+                    .physical_location(sarif::PhysicalLocationBuilder::default()
+                        .artifact_location(
+                            sarif::ArtifactLocationBuilder::default()
+                                .uri(msg.context.as_ref().map(|context| match context {
+                                    Context::File(context) => context.path.to_str().unwrap_or(""),
+                                    _ => "",
+                                }).unwrap_or(""))
+                                .build().map_err(|e| { error_to_io_error(e.into()) })?
+                        )
+                        .region(
+                            sarif::RegionBuilder::default()
+                                .start_line(line_num as i64)
+                                .start_column(column_start as i64)
+                                .end_column(column_end as i64)
+                                .build().map_err(|e| { error_to_io_error(e.into()) })?
+                        )
+                        .build().map_err(|e| { error_to_io_error(e.into()) })?)
+                    .build().map_err(|e| { error_to_io_error(e.into()) })?;
+
+
+                let result = sarif::ResultBuilder::default()
+                    .level(sarif::ResultLevel::Error.to_string())
+                    .message(sarif::MessageBuilder::default()
+                        .markdown(message)
+                        .build().map_err(|e| { error_to_io_error(e.into()) })?)
+                    .locations(vec![location])
+                    .build().map_err(|e| { error_to_io_error(e.into()) })?;
+
```

Comment:
One way to simplify these `map_err`s is to pull the logic out into a helper that returns the error from sarif.

Reply:
This isn't done.  In case there was confusion, this was about splitting it out like
```rust


impl Report for PrintSarif {
    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
        report_sarif(msg).map_err(error_to_io_error)
    }
}

fn report_sarif(&self, msg: Message<'_>) -> Result<(), _> {
    // ...
}
```

Reply:
Meaning implement this only for sarif, not extract a trait that requires context?

Reply:
Unsure what is meant by either option of the question.

I'm just referring to extracting the function body that has a sarif-focued error type so you can use `?` without having to `map_err` every time, only doing it in the wraper call that handles the trait impl.

Reply:
I see what you mean. I'll update the code now.

Reply:
updated

Reply:
ok